### ISSUE TITLE
sshcollector migration fails for devices not existing or not created as pseudo

### ISSUE THREAD

**Issue by hannut** [netdisco/netdisco#587]
migrated old sshcollector to new arpnip -based solution with instructions from here: https://github.com/netdisco/netdisco/wiki/sshcollector-Deprecation

This is a commercial linux-based firewall which has worked with the old collector.

$ ~/bin/netdisco-do arpnip -d nn.nn.nn.nn
[14030] 2019-05-22 07:12:54  info App::Netdisco version 2.042008 loaded.
[14030] 2019-05-22 07:12:55  info arpnip: [nn.nn.nn.nn] started at Wed May 22 10:12:55 2019
[14030] 2019-05-22 07:12:55  info arpnip: finished at Wed May 22 10:12:55 2019
[14030] 2019-05-22 07:12:55  info arpnip: status info: arpnip skipped: nn.nn.nn.nn has no layer 3 capability

ps. sorry for my English, writing in foreign language :)

**Comment by rc9000** [comment]
Yes the new version requires the device to advertise layer 3 capabilities via the SNMP sysServices object. Can you configure the SNMP daemon on this machine? Here is how you'd do it for standard UCD snmpd, which you are likely to run:

 * http://www.net-snmp.org/docs/man/snmpd.conf.html search for "sysServices"
 * general description of OID http://cric.grenoble.cnrs.fr/Administrateurs/Outils/MIBS/?oid=1.3.6.1.2.1.1.7

Once you have configured snmpd, please rediscover the device, then arpnip should run for it.

**Comment by hannut** [comment]
Thanks for the reply. Problem is that this is a commercial firewall so all the nice Linux-knobs are glued in place by the vendor. Is there a way to manually add L3 to the device on the netdisco-side?

**Comment by rc9000** [comment]
Yes you can just set the discovered value directly in the database. Unfortunately it will be overwritten at discovery, so you'd need to use a cronjob like e.g.

    netdisco-do psql -e "update device set layers = '0000110' where host(ip) in  ('x.x.x.x', 'x.x.x.y', ...all your fw where this applies)" 

where you set the layers you want activated to 1 - this starts from the right so you'll want 0000110 at least to get Netdisco macsuck and arpnip on it. You need to run this after every discovery. Of course the where could also refer to vendor, model etc.

There are some other options like using a pseudo device or create the above sql as a trigger on the device table, depending a bit on your requirements and familiarity with poking around in Postgres and Netdisco.


PS sorry I don't have access to a Netdisco right now, so the command is untested.

**Comment by inphobia** [comment]
what snmp::info class is used for your device?
`netdisco-do arpnip -d nn.nn.nn.nn -DI`
`netdisco-do show -e specify -d nn.nn.nn.nn -DI`
(you can delete any info you think is private)

there are a few more, perhaps easier options i can think of:
* overwrite the layers() function in snmp::info for you device type, but that will need to be done each time you upgrade snmp::info
* create a new snmp::info class for your device, but that seems a bit overkill
* add logic to the snmp::info class to overwrite layers() just for your device type.
* add a knob to driver: cli to force arpnip even if sysServices don't match.

i'm kinda a fan of the last one, we have sshcollector for devices that have issues with snmp, so being able to force ssh seems like a decent way out. this will require also some work. the question then is, what do you do with the layers field in the device table?

we can add l3 capabilities when arpnip runs, but then it will just get overwritten again at the next discover. but not setting l3 might lead to instances where netdisco ignores the device's arp info for certain functions.

pseudo devices are an option but those also have different code paths sometimes in netdisco.

@hannut 
i think for now you can have a look at https://github.com/netdisco/netdisco/wiki/Configuration#use_legacy_sshcollector to get the old collector going again until we've come up with a solution

@netdisco/netdisco-admins @netdisco/netdisco-developers 
bottom line, imo an option to force arpnip with driver: cli and some kind of "don't let discover overwrite layers" (perhaps add a user_layers field do the the device table that discover won't touch and then join that with discover layers?) seems interesting but requires quite some work.

i've been playing with the idea of some kind of user overwrite for whatever netdisco discovers. i know there's snmp_set for some modules but that support is sparse.

actual use case: one of our providers allows snmp read access but not write. we would like to use the "location" info but not the one provided.

a device_overwrite table might be an idea, but then where to draw the line. do we want overwrite for all device_* tables? it would give things like has manual topo a decent place to live.

thoughts?

**Comment by ollyg** [comment]
fixed in
[master 2897eda6] #587 #561 update pseudo devices to better support ssh arpnip
